cloud/ec2_fleet.go

```diff
@@ -20,61 +20,62 @@ import (
+	if subnets, ok := typeCache.instanceTypeToSubnets[instanceRegion]; ok {
+		return subnets, nil
+	}
+
+	supportingAZs, err := c.getAZs(ctx, settings, client, instanceRegion)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't get supporting AZs")
 	}
 
-	subnets := make([]evergreen.Subnet, 0, len(c.instanceTypeToSubnets[instanceType]))
-	for _, subnet := range c.instanceTypeToSubnets[instanceType] {
-		if AztoRegion(subnet.AZ) == region {
+	subnets := make([]evergreen.Subnet, 0, len(supportingAZs))
+	for _, subnet := range settings.Providers.AWS.Subnets {
+		if utility.StringSliceContains(supportingAZs, subnet.AZ) {
 			subnets = append(subnets, subnet)
 		}
 	}
+	c.instanceTypeToSubnets[instanceRegion] = subnets
+
 	return subnets, nil
 }
 
-func (c *instanceTypeSubnetCache) Build(ctx context.Context, settings *evergreen.Settings, client AWSClient) error {
-	subnets := settings.Providers.AWS.Subnets
-	if len(subnets) == 0 {
-		return errors.New("no AWS subnets were configured")
+func (c *instanceTypeSubnetCache) getAZs(ctx context.Context, settings *evergreen.Settings, client AWSClient, instanceRegion instanceRegionPair) ([]string, error) {
+	output, err := client.DescribeInstanceTypeOfferings(ctx, &ec2.DescribeInstanceTypeOfferingsInput{
+		LocationType: aws.String(ec2.LocationTypeAvailabilityZone),
+		Filters: []*ec2.Filter{
+			{Name: aws.String("instance-type"), Values: []*string{aws.String(instanceRegion.instanceType)}},
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "can't get instance types for '%s' in '%s'", instanceRegion.instanceType, instanceRegion.region)
 	}
-
-	for _, subnet := range subnets {
-		output, err := client.DescribeInstanceTypeOfferings(ctx, &ec2.DescribeInstanceTypeOfferingsInput{
-			LocationType: aws.String(ec2.LocationTypeAvailabilityZone),
-			Filters: []*ec2.Filter{
-				{Name: aws.String("location"), Values: []*string{aws.String(subnet.AZ)}},
-			},
-		})
-		if err != nil {
-			return errors.Wrapf(err, "can't get instance types for '%s'", subnet.AZ)
-		}
-		if output == nil {
-			return errors.Errorf("DescribeInstanceTypeOfferings returned nil output for AZ '%s'", subnet.AZ)
-		}
-		for _, offering := range output.InstanceTypeOfferings {
-			if offering != nil && offering.InstanceType != nil {
-				c.instanceTypeToSubnets[*offering.InstanceType] = append(c.instanceTypeToSubnets[*offering.InstanceType], subnet)
-			}
+	if output == nil {
+		return nil, errors.Errorf("DescribeInstanceTypeOfferings returned nil output for AZ '%s' in '%s'", instanceRegion.instanceType, instanceRegion.region)
+	}
+	supportingAZs := make([]string, 0, len(output.InstanceTypeOfferings))
+	for _, offering := range output.InstanceTypeOfferings {
+		if offering != nil && offering.Location != nil {
+			supportingAZs = append(supportingAZs, *offering.Location)
```

Comment:
Don't we only want to include availability zones that are in the requested region? It doesn't look like we do anything with instanceRegion.region

Reply:
The region isn't used explicitly in the call, but the response will only include the AZs in the client's region.

For example, if we call it with a client configured with the us-east-1 region and get the AZs there I want that the next time we call it with a client configured for us-east-2 we should know that we haven't called it for this region yet and make the API call.

Reply:
Oh I see what you mean. Can you put a comment to that effect where DescribeInstanceTypeOfferings is called then, so it's clear that region is implicitly being considered?